matching-core/src/main/java/com/graphhopper/matching/GPXExtension.java

```
@@ -17,36 +17,95 @@
-    final GPXEntry entry;
-    final QueryResult queryResult;
-    final int gpxListIndex;
+    private final GPXEntry entry;
+    private final QueryResult queryResult;
+    private final boolean isDirected;
+    private final EdgeIteratorState incomingVirtualEdge;
+    private final EdgeIteratorState outgoingVirtualEdge;
 
-    public GPXExtension(GPXEntry entry, QueryResult queryResult, int gpxListIndex) {
+    /**
+     * Creates an undirected candidate for a real node.
+     */
+    public GPXExtension(GPXEntry entry, QueryResult queryResult) {
         this.entry = entry;
         this.queryResult = queryResult;
-        this.gpxListIndex = gpxListIndex;
+        this.isDirected = false;
+        this.incomingVirtualEdge = null;
+        this.outgoingVirtualEdge = null;
     }
 
-    @Override
-    public String toString() {
-        return "entry:" + entry
-                + ", query distance:" + queryResult.getQueryDistance()
-                + ", gpxListIndex:" + gpxListIndex;
+    /**
+     * Creates a directed candidate for a virtual node.
+     */
+    public GPXExtension(GPXEntry entry, QueryResult queryResult,
+                        VirtualEdgeIteratorState incomingVirtualEdge,
+                        VirtualEdgeIteratorState outgoingVirtualEdge) {
+        this.entry = entry;
+        this.queryResult = queryResult;
+        this.isDirected = true;
+        this.incomingVirtualEdge = incomingVirtualEdge;
+        this.outgoingVirtualEdge = outgoingVirtualEdge;
+    }
+
+    public GPXEntry getEntry() {
+        return entry;
     }
 
     public QueryResult getQueryResult() {
-        return this.queryResult;
+        return queryResult;
     }
 
-    public GPXEntry getEntry() {
-        return entry;
+    /**
+     * Returns true if the snapped point is a virtual node, otherwise returns false.
+     */
+    public boolean isDirected() {
+        return isDirected;
+    }
+
+    /**
+     * Returns null if this GPXExtension is not directed.
```

Should we instead throw an exception to enforce that `isDirected` is always called before?